modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/recovery/RecoveryIoImp.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.ignite.lang.IgniteBiInClosure;
+import org.apache.ignite.lang.IgniteInClosure;
+import org.apache.ignite.plugin.extensions.communication.Message;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.GridTopic.TOPIC_RECOVERY;
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.SYSTEM_POOL;
+
+/**
+ *
+ */
+public class RecoveryIoImp implements RecoveryIo, GridMessageListener, DiscoveryEventListener {
+    /** */
+    private final GridKernalContext ctx;
+
+    /** */
+    private final ConsistentIdMapper idMapper;
+
+    /** */
+    private volatile IgniteBiInClosure<String, Message> msgHandler;
+
+    /** */
+    private volatile IgniteInClosure<String> nodeLeftHandler;
+
+    /** */
+    private final IgniteLogger log;
+
+    /**
+     *
+     */
+    public RecoveryIoImp(GridKernalContext ctx, @Nullable IgniteLogger log) {
+        idMapper = new ConsistentIdMapper(ctx.discovery());
+
+        this.log = log;
+        this.ctx = ctx;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void send(String constId, Message msg) throws IgniteCheckedException {
+        UUID nodeId = idMapper.mapToUUID(constId);
+
+        assert nodeId != null;
+
+        ctx.io().sendToGridTopic(nodeId, TOPIC_RECOVERY, msg, SYSTEM_POOL);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void receive(final IgniteBiInClosure<String, Message> handler) {
+        msgHandler = handler;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onNodeLeft(final IgniteInClosure<String> handler) {
+        nodeLeftHandler = handler;
+    }
+
+    /** {@inheritDoc} */
+    @Override public List<String> constIds(AffinityTopologyVersion ver) {
+        List<String> constIds = new ArrayList<>();
+
+        for (Object constId : idMapper.mapToConsistentId(ver))
+            constIds.add(constId.toString());
+
+        return constIds;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String localNodeConsistentId() throws IgniteCheckedException {
+        try {
+            return ctx.pdsFolderResolver().resolveFolders().consistentId().toString();
+        }
+        catch (IgniteCheckedException e) {
+            U.error(log, "Fail resolve consistent id.", e);
+
+            new Thread(
+                new Runnable() {
+                    @Override public void run() {
+                        G.stop(true);
+                    }
+                }
+            ).start();
+
+            throw e;
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onMessage(UUID nodeId, Object msg, byte plc) {
+        IgniteBiInClosure<String, Message> handler = msgHandler;
+
+        if (handler != null && msg instanceof Message) {
+            String constId = idMapper.mapToConsistentId(
+                ctx.discovery().topologyVersionEx(), nodeId
+            ).toString();
+
+            handler.apply(constId, (Message)msg);
+        }
```

Comment:
We should not silently drop a message. If a listener is supposed to be always installed before a first message is sent, we should at least print a warning.